main/lsp/ConvertToSingletonClassMethod.cc

```diff
@@ -0,0 +1,208 @@
+    ENFORCE(insertSelfRange != nullptr);
+    auto insertParamRange = Range::fromLoc(gs, insertParamLoc);
+    ENFORCE(insertParamRange != nullptr);
+
+    auto uri = config.fileRef2Uri(gs, definition.termLoc.file());
+    auto tdi = make_unique<VersionedTextDocumentIdentifier>(move(uri), JSONNullObject());
+    vector<unique_ptr<TextEdit>> edits;
+    edits.emplace_back(make_unique<TextEdit>(move(insertSelfRange), "self."));
+    if (needsParens) {
+        edits.emplace_back(make_unique<TextEdit>(move(insertParamRange), "(this)"));
+    } else if (definition.symbol.data(gs)->arguments.empty()) {
+        edits.emplace_back(make_unique<TextEdit>(move(insertParamRange), "this"));
+    } else {
+        edits.emplace_back(make_unique<TextEdit>(move(insertParamRange), "this, "));
+    }
+
+    if (definition.symbol.data(gs)->hasSig()) {
+        auto trees = typechecker.getResolved({file});
+        ENFORCE(!trees.empty());
+        auto &rootTree = trees[0].tree;
+
+        auto ctx = core::Context(gs, core::Symbols::root(), file);
+        auto queryLoc = definition.termLoc.copyWithZeroLength();
+        auto parsedSig = sig_finder::SigFinder::findSignature(ctx, rootTree, queryLoc);
+        if (parsedSig.has_value()) {
+            if (!parsedSig->argTypes.empty()) {
+                auto firstArgLoc = parsedSig->argTypes[0].nameLoc;
+                auto insertSigParamRange = Range::fromLoc(gs, firstArgLoc.adjustLen(gs, 0, 0));
+                auto sigParamText = fmt::format("this: {}, ", definition.symbol.data(gs)->owner.show(gs));
+                edits.emplace_back(make_unique<TextEdit>(move(insertSigParamRange), move(sigParamText)));
+            } else if (parsedSig->returnsLoc.exists()) {
+                auto insertSigParamsRange = Range::fromLoc(gs, parsedSig->returnsLoc.adjustLen(gs, 0, 0));
+                auto sigParamText = fmt::format("params(this: {}).", definition.symbol.data(gs)->owner.show(gs));
+                edits.emplace_back(make_unique<TextEdit>(move(insertSigParamsRange), move(sigParamText)));
+            }
+        }
+    }
+
+    return make_unique<TextDocumentEdit>(move(tdi), move(edits));
+}
+
+class MethodCallSiteRewriter : public AbstractRewriter {
+    core::MethodRef method;
+    core::ClassOrModuleRef owner;
+
+public:
+    MethodCallSiteRewriter(const core::GlobalState &gs, const LSPConfiguration &config, core::MethodRef method)
+        : AbstractRewriter(gs, config), method(method), owner(method.data(gs)->owner) {}
+    ~MethodCallSiteRewriter() {}
+
+    void rename(unique_ptr<core::lsp::QueryResponse> &response, const core::SymbolRef originalSymbol) override {
+        if (invalid) {
+            return;
+        }
+
+        auto sendResp = response->isSend();
+        if (sendResp == nullptr || !sendResp->receiverLocOffsets.exists()) {
+            return;
+        }
+
+        if (sendResp->callerSideName == core::Names::callWithSplat() ||
```

Comment:
When I don't do any special handling, I get this:

```
A.<call-with-splat>(A.new.takes_block(0, *args), A.new.takes_block(0, *args))
```

looking at that, it tells me that if I were to try to make this work for splats too, it would run into the same problems I ran into when trying to get blockpasses to work (I spent a couple hours on it and nothing quite worked).

Personally, I think we should overhaul how these argument passing things work, in a future change.